ext/drreg/drreg.h

```diff
@@ -324,6 +324,15 @@ typedef enum {
      * to be handled by the user manually, usually via drreg_restore_all().
      */
     DRREG_USER_RESTORES_AT_BB_END = 0x004,
+
+    /**
+     * Turns on stricter logic to find free register spill slots. This makes sure
+     * that a slot wasn't used to spill some register value in prior instrumentation
+     * passes. An example usage is in drx_expand_scatter_gather which is used in
+     * the app2app pass and requires spilling of registers to slots that may
+     * conflict with slots used during later instrumentation passes.
+     */
+    DRREG_HANDLE_MULTI_PHASE_SLOT_RESERVATIONS = 0x008,
```

Comment:
Should this be on by default, instead of off?  It has a performance cost to walk the ilist every single time, and there could be a problem if code were somewhere inserted beyond the current instr: there are many cases of postinserts, but not by drreg code with unique notes.

I think the biggest issues would be instru2instru phases: but they already have conflict issues and might want this on anyway?  Although actually bb_props is set to 0 in the insertion phase?  **Should that be moved to a late instru2instru event?**

Reply:
> Should this be on by default, instead of off? It has a performance cost to walk the ilist every single time

Do you expect multi-phase (or multiple clients in the same phase) use of drreg to be very frequent? If yes, then I agree we should keep it on by default. Otherwise, I'm concerned about the unnecessary overhead. We can keep the note-keeping on by default, and let clients add this flag if they use drreg outside instrumentation phase (like in instru2instru phase). To catch some of these more common cases, we can maybe add the flag ourselves in drreg if we see a usage in these phases.

> Although actually bb_props is set to 0 in the insertion phase? Should that be moved to a late instru2instru event?

What if the client didn't register any phase after insertion?

Reply:
> Do you expect multi-phase (or multiple clients in the same phase) use of drreg to be very frequent? 

Every address tracing client is going to use drx_expand_scatter_gather, which uses drreg in the app2app phase, so such usage will not be uncommon.

> > Although actually bb_props is set to 0 in the insertion phase? Should that be moved to a late instru2instru event?
>
> What if the client didn't register any phase after insertion?

Not sure I follow the question: there would be no harm then in moving the set to 0 later.

Reply:
>> Do you expect multi-phase (or multiple clients in the same phase) use of drreg to be very frequent?
>
> Every address tracing client is going to use drx_expand_scatter_gather, which uses drreg in the app2app phase, so such usage will not be uncommon.

The flag is already set when `drx_expand_scatter_gather` is invoked. So that is covered.

My comment was about other user-owned DR clients where scatter/gather instrs may not be expected or not important. I don't suppose we need to do anything proactively for those clients right? Or do you mean we should proactively set the flag everywhere, anticipating uses in multiple passes like instru2instru?

>>> Although actually bb_props is set to 0 in the insertion phase? Should that be moved to a late instru2instru event?
>>
>> What if the client didn't register any phase after insertion?
>
> Not sure I follow the question: there would be no harm then in moving the set to 0 later.

If there's no instru2instru event registered by the user (or if there is one but the user doesn't invoke drreg in it), then we won't be able to reset `bb_props` to zero. Do you mean that drreg should register a new low-priority instru2instru event that will get invoked at the very end, which resets `bb_props`?

Reply:
> The flag is already set when drx_expand_scatter_gather is invoked. So that is covered.
>
> My comment was about other user-owned DR clients where scatter/gather instrs may not be expected or not important. I > don't suppose we need to do anything proactively for those clients right? Or do you mean we should proactively set the > flag everywhere, anticipating uses in multiple passes like instru2instru?

If the cost is not very high it seems better to just have it on and avoid subtle hard-to-debug problems in future clients.  OTOH as you say this app2app scatter-gather case is covered already.
If we leave it off by default: could you add clear comments to the drreg docs so that anyone using it in app2app or instru2instru knows to go set this flag?

> Do you mean that drreg should register a new low-priority instru2instru event that will get invoked at the very end, which resets bb_props?

Yes, drreg would register its own event: it couldn't use an event registered by other code even if it wanted to as it has no idea where other event callbacks are.

Reply:
I'm now saving data about spill slot usage even when the option isn't set. The new spill slot selection logic itself however is still behind the option .Added documentation to `drreg.dox` and `release.dox` about the need to set this option for multi-phase usage of drreg.

> Yes, drreg would register its own event: it couldn't use an event registered by other code even if it wanted to as it has no idea where other event callbacks are.

Done

Reply:
> I'm now saving data about spill slot usage even when the option isn't set

Should we consider removing the option then for simplicity and ease of testing?

This data would be useful for the restore_state event: if we had #3801.

Reply:
> Should we consider removing the option then for simplicity and ease of testing?

The option is still required to enable the stricter, more expensive spill slot selection logic. The only change I made was to save spill slot usage data regardless of the option, so that the logic works even if the option is turned on after first drreg use.